compiler+runtime/src/jank/clojure/core.jank

```diff
@@ -3447,34 +3447,66 @@
        (def ~name ~expr))))
 
 ;; Case.
-(defn- shift-mask [shift mask x]
-  (-> x (bit-shift-right shift) (bit-and mask)))
-
 (def ^:private max-mask-bits 13)
 (def ^:private max-switch-table-size (bit-shift-left 1 max-mask-bits))
 (def ^:private int-min-value clojure.core-native/int-min)
 (def ^:private int-max-value clojure.core-native/int-max)
+(def ^:private int32-min-value clojure.core-native/int32-min)
+(def ^:private int32-max-value clojure.core-native/int32-max)
+
+(defn- shift-mask [shift mask x]
+  (-> x (bit-shift-right shift) (bit-and mask)))
+
+(defn- case-hash
+  "Returns the input if it is within the range of a 32-bit signed integer, otherwise returns the hash of the input."
+  [input]
+  (if (and (integer? input)
+           (>= input int32-min-value)
+           (<= input int32-max-value))
+    input
+    (hash input)))
 
 (defn- maybe-min-hash
   "Takes a collection of hashes and returns [shift mask] or nil if none found"
   [hashes]
   (first
-    (filter (fn [[s m]]
-              (apply distinct? (map #(shift-mask s m %) hashes)))
-            (for [mask (map #(dec (bit-shift-left 1 %)) (range 1 (inc max-mask-bits)))
-                  shift (range 0 31)]
-              [shift mask]))))
-
-(defn- case-map
+   (filter
+    (fn [[s m]]
+      (apply distinct? (map #(shift-mask s m %) hashes)))
+    (for [mask
+          (map #(dec (bit-shift-left 1 %)) (range 1 (inc max-mask-bits)))
+          shift
+          (range 0 31)]
+      [shift mask]))))
+
+(defn- case-map-1
+  [expr-sym default case-f test-f tests thens]
+  (into (sorted-map)
+        (zipmap (map case-f tests)
+                (map
+                 (fn [key then]
+                   (let [key
+                         (if (symbol? key)
+                           (list 'quote key)
+                           key)]
+                     `(condp = ~expr-sym ~key ~then
+                       ~default)))
+                 (map test-f tests)
+                 thens))))
+
+(defn- case-map-2
   "Transforms a sequence of test constants and a corresponding sequence of then
    expressions into a sorted map to be consumed by case*. The form of the map
    entries are {(case-f test) [(test-f test) then]}."
   [case-f test-f tests thens]
   (into (sorted-map)
-        (zipmap (map case-f tests)
-                (map vector
-                     (map test-f tests)
-                     thens))))
+        (zipmap (map case-f tests) thens)))
+
+(defn- case-map
+  [expr-sym default case-f test-f tests thens skip-check]
+  (if (empty? skip-check)
+    (case-map-1 expr-sym default case-f test-f tests thens)
+    (case-map-2 case-f test-f tests thens)))
```

Comment:
I think we can do better with naming than `-1` and `-2` suffixes. What's the difference between this and why is it needed? That also seems worth a comment, above this `if`.